fix: Better error for unzip failure

Sudha247 opened the pull request:
Fixes #11548

Currently the archive driver tries to use `bsdtar` and `tar` to unzip a zip file if `unzip` is not present in the system. This may or may not work depending on the specific implementation. Concretely, if the `tar` is a variant of `bsdtar` it can unpack zip files whearas `tar` cannot. We check that and return an error message that `unzip` is expected and not present.

Sudha247 committed:
(fix): Better error for unzip failure

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Sudha247 committed:
Address review comments

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Leonidas-from-XIV commented on src/dune_pkg/archive_driver.ml:
```diff
@@ -26,38 +26,50 @@ let make_zip_args ~archive ~target_in_temp =
+  let bin_path =
+    match which bin_name with
+    | Some bin -> bin
+    | None -> Dune_engine.Utils.program_not_found "unzip" ~loc:None
```
This is probably wrong, if it searches for `tar` or `bsdtar` it shouldn't fail with an error that `unzip` can't be found.

```suggestion
    | None -> Dune_engine.Utils.program_not_found bin_name ~loc:None
```

Sudha247 replied:
Hm, I'm not so sure about this. The reason we're looking for `bsdtar` or `tar` is because `unzip` doesn't exist, and `unzip` is the definitive program to extract zip files, while `tar` may or may not extract zip files. I think the error message should be clear here that the failure is because a zip file is not being extracted. Should we write a custom error message here?

Leonidas-from-XIV reviewed (changes requested):
Looks pretty sensible! I found one thing that's probably an actual issue.

Leonidas-from-XIV reviewed (approved):
Looks good to me. It would be nice to add a cram test, however removing tools from the `PATH` on Unix is rather tedious (as one would need to iterate over the whole `PATH`, symlink those that should stay in the path and skip those that should be excluded).

Alizter commented:
We have such a test for `curl`, we could reuse that setup.

Leonidas-from-XIV commented:
Hmm, yes, looks like we don't actually need many binaries to exist in the `PATH`. So a fake `PATH` folder is fairly easy to create, only needs `g(un)zip`, `tar` `unzip` etc.

Sudha247 committed:
add cram test for the case when zip extraction fails

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Leonidas-from-XIV reviewed (changes requested):
I think there's some issues in the cram test that would be nice to fix to have a more comprehensive (and faster) coverage of decompression tool testing.

Sudha247 committed:
test various scenarios in zip extract fail test

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Sudha247 committed:
archive: expand the error message to list all the binaries tried

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Sudha247 committed:
test: make aliases so the test doesn't depend on binaries.

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Alizter commented on src/dune_pkg/archive_driver.ml:
```diff
@@ -26,38 +26,56 @@ let make_zip_args ~archive ~target_in_temp =
+           Fiber.return
+           @@ User_error.raise
+                [ Pp.enumerate
+                    [ "No program found to extract zip file. Tried"
```
```suggestion
                [ "No program found to extract zip file. Tried:"
                ; Pp.enumerate
                  [
```

Sudha247 committed:
Update test to use fake binaries as the binaries are not available in some platforms

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Sudha247 committed:
Update test to use fake binaries as the binaries are not available in some platforms

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Alizter commented:
This issue is relevant:
- https://github.com/ocaml/dune/issues/10123

This issue also makes all the work being done here pointless if we ever go down that route:
- https://github.com/ocaml/dune/issues/10127

I don't think this PR and it's coverage need to be perfect.

Sudha247 committed:
update cram test to include more cases and fake binaries that work

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

Leonidas-from-XIV reviewed (approved):
Pretty nice. I think some debug prints can be removed but otherwise its good to go.

Sudha247 committed:
remove leftover debug prints

Signed-off-by: Sudha Parimala <sudharg247@gmail.com>

shonfeder committed:
Fix cram tests on MacOS

It seems we need to run the command with the PATH=... prefix in a
subshell, or else the PATH can pollute subsequent commands.

Signed-off-by: Shon Feder <shon.feder@gmail.com>

Alizter commented:
@shonfeder I would like to give this another pass tomorrow if you don't mind.

rgrinberg commented on src/dune_pkg/archive_driver.ml:
```diff
@@ -26,38 +26,51 @@ let make_zip_args ~archive ~target_in_temp =
+    let+ output, _error = Process.run_capture ~display:Quiet Return bin [ "--version" ] in
+    let re = Re.compile (Re.str "bsdtar") in
+    if Re.execp re output then Some bin else None
+  | None -> Fiber.return None
```
Do you mind putting this clause the `Some`?

shonfeder committed:
Merge branch 'main' into unzip-archive
